Paper: Software Engineering as Research Method

sbenthall opened the pull request:
@MridulS and I expect to be making some additions over the course of the review period.

sbenthall committed:
stub for paper

sbenthall committed:
references

sbenthall committed:
lit review and econ-ark introduction

sbenthall committed:
adding stuff about roles

MridulS committed:
add some stuff

sbenthall committed:
Merge pull request #1 from MridulS/2020_econ_ark

add some stuff

sbenthall committed:
finishing roles section, adding paragraph to discussion section

sbenthall committed:
adding abstract, user roles breakdown, and discussion

MridulS committed:
initial changes

MridulS committed:
add first draft text

sbenthall committed:
Merge pull request #2 from MridulS/2020_econ_ark

add first draft text

sbenthall committed:
some small edits

sbenthall committed:
adding research engineer references

sbenthall committed:
adding note about the COVID dashboard

sbenthall committed:
adding example problem

sbenthall committed:
adding note about NSTF

sbenthall committed:
edits

sbenthall commented:
Are there reviewers assigned for this article?

deniederhut commented:
Yes -- reviewers were assigned to your paper on 31 May.

davclark commented:
@sbenthall shared a link to this submission, so while I'm not a randomly assigned reviewer, I'll add a few notes in case they are useful.

First, I think this paper situates a useful case study within a broader high-impact domain (economic prediction). The specific case is almost certainly novel (as @sbenthall is a member of the research project), and to my knowlege the general framing is likewise tying together major long-term trends in science - tracing back to the introduction of the use of computers in science, as well as recent thinking in the philosophy of science - in a way that I've not seen in the Scientific Python community.

As a fan of Papert, I find myself yearning for a clearer statement of a thesis, and I think it's a claim that Papert made in *Mindstorms* - that academia desperately needs to include individuals who can make increasingly siloed knowledge intelligible to a broader community. But, I wouldn't push hard for this if the authors are happy with how they've organized their main points.

Also - hi @deniederhut!

mepa commented:
Hi @sbenthall. None of the assigned reviewers have responded to our invitation to review so I will be stepping in to help review your submission. This is a topic that truly interests me and I look forward to reading in detail. I'll complete the review this weekend. Please note that @deniederhut will soon announce an extension of the initial review deadline, which was today, to give reviewers a bit more time.

@davclark, thank you for jumping in with your comments!

deniederhut commented:
Hey @sbenthall ! Thanks to some awesome behind the scenes work by @stargaser, we now have experimental support for publishing papers with ORCIDs. If you have one and would like to add it to your paper, it's as simple as adding an `:orcid:` tag under your name in the paper header. [You can see an example of how to do this in the `00_vanderwalt` example paper](https://raw.githubusercontent.com/scipy-conference/scipy_proceedings/2020/papers/00_vanderwalt/00_vanderwalt.rst). If you don't have one, no worries! You can still publish your paper as is.

sbenthall committed:
reorganizing some materials

sbenthall committed:
adding orcid and corresponding author tag

sbenthall commented:
@mepa Thank you for reviewing!

@davclark Thanks for your comment! My copy of _Mindstorms_ is locked away at the moment, but I will try to clarify his thesis in a later draft. The constructionism vs. computational thinking frame is taken from Barba's piece here:

https://lorenabarba.com/blog/computational-thinking-i-do-not-think-it-means-what-you-think-it-means/

which is cited but I wonder if it could be better worked in. Maybe a more complete piece analyzing Papert's connection to scientific programming would be a good topic for a different, standalone piece.

@deniederhut I've added my ORCID. Thanks for the heads up. @MridulS might add his too.

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,625 @@
+it was not long before Papert's Logo tool was also
+used in scientific education, teaching students not just
+about the abstract mathematical sphere, but about the
+physical world. :cite:`resnick1990lego`.
```
`world.` <-- period not needed

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,625 @@
+physical world. :cite:`resnick1990lego`.
+The legacy of Logo is alive and well in NetLogo :cite:`tisue2004netlogo`,
+which is used by both students and researchers alike in the study of
+complex and agent-based systems, and the Python ABM toolkit
```
- `Python agent-based modeling (ABM) toolkit`?
- It may be helpful to the uninitiated reader to give a concrete example of a "complex" or "agent-based" system that NetLogo is used to study. 
- `and the Python` ->  `and *in* the Python`

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,625 @@
+   This gap will not be closed until software engineering
+   is considered a core skill for the discipline.
+   Software engineering should be reconceived as a research
+   method. 
```
I think it becomes more clear later in the text, but would it be possible here in the abstract to clarify (1) what are the opportunities being missed by members of the three roles when there is a lack in software engineering skills, and (2) why is depending solely on research software engineers insufficient (if I'm interpreting that correctly)?

mepa replied:
Thanks for making the changes! The abstract sounds great, with the minor correction of the sentence noted above.

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,625 @@
+modeling :cite:`hommes2006heterogeneous`,
+which became a research priority
+after the 2008 financial crisis revealed the weaknesses in the
+then-dominant representative agent (RA) based paradigm [#]
```
`paradigm [#]_` (missing the trailing underscore so it prints as "[#]" in the pdf)

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,625 @@
+Econ-ARK lies roughly in the Papertian educational tradition,
+similar to other agent-based modeling software such as
+NetLogo :cite:`tisue2004netlogo` and Mesa :cite:`masad2015mesa`.
+However, in Econ-ARK models, agents that optimize their
```
`agents optimize their` (remove "that")

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,625 @@
+In this respect, Econ-ARK has some characteristics of a
+reinforcement learning or "AI" toolkit.
+
+**Example**. An paradigmatic, simple example of the kind
```
`An` --> `A`

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,625 @@
+where :math:`U` is a utility function, :math:`\rho` is a
+coefficient of risk aversion, :math:`c` is the
+amount of resources the agent chooses to consume in each
+period, :math:`m` are the market resources available to
```
`is the amount of market resources`?

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,625 @@
+salience of the finding.
+This entails that the success of Econ-ARK
+will imply a practical change to the research field: students will
+study models that have been published by researchers in Python
```
I would move `in Python` to clarify meaning: `have been published in Python by researchers`

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,625 @@
+study models that have been published by researchers in Python
+in order to learn insights about the economy. This blending of roles,
+between researchers, students, and software engineers,
+complicates the software architecture of the toolkit
```
This left me wondering how the blending of roles "complicates the software architecture of the toolkit", so it may be nice to add a few words summarizing how, or a phrase like `as described in the next section`.

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,625 @@
+nebulously defined because while the question of how to
+implement a class of economic models efficiently and robustly
+in Python is a "research question" in its own right, these
+kinds of contributions are rarely considered first-order research contributions. Researchers work within a complex field of economic capital incentives (such as
```
I agree this is a problem, and not only in economics!

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,625 @@
+of an adequate computing environment that does not distract
+them from the course materials. Technical solutions
+have been developed for both hurdles. `nbgrader` enables
+the creation of assignments with Jupyter notebooks. :cite:`hamrick2016creatin` :cite:`blank2019nbgrader`
```
- `hamrick2016creating` (missing the "g", so it's rendered in the pdf as `[?]`)
- I think the period should go after the citations.

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,625 @@
+via GitHub. The workflow patterns of collaboratively
+developing software using GitHub and managing release cycles
+are not part of any conventional Economics curriculum,
+and yet these are researcher increasingly need to learn and
```
`and yet researchers increasingly need to learn and use these in order to`

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,625 @@
+They must be learned informally by researchers who are
+not incentivized to do so, or they must be hired from
+an external talent pool trained in other disciplines
+or at another workplace.
```
This is a well-stated point.

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,625 @@
+JupyterHub has been deployed to allow students to get around the
+hardware limitations of their laptops and the difficulties of
+setting up a local coding environment. :cite:`kim_2018`
+Noteably, both technical solutions, which have been developed only
```
`Notably` (remove the "e")

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,625 @@
+the creation of assignments with Jupyter notebooks. :cite:`hamrick2016creatin` :cite:`blank2019nbgrader`
+JupyterHub has been deployed to allow students to get around the
+hardware limitations of their laptops and the difficulties of
+setting up a local coding environment. :cite:`kim_2018`
```
Period goes after citation.

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,625 @@
+professor, who teaches more students, which is necessary
+for the autonomy of economics as a field of knowledge.
+If at every point in the process--even at the point where
+new discoveries are integrated into the core software library--
```
Leave spaces around `--` in both cases? I'm not sure what is proper punctuation, but consistency between the two would be good.

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,625 @@
+
+Case Study: Econ-ARK infrastructure
+------------------------------------------
+The Econ-ARK infrastructure is built around creating a sustainable community with respect to various different use cases and the challenges of creating sustainable scientific software. We discuss some of the challenges of bridging work across user roles of Researchers, Publishers, Pedagogy and Software engineering. This is discussed from the point of view of economics research and tools in the Sc ...
```
`Researchers, Publishers, Pedagogues, and Software Engineers` (since these are "roles")

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,625 @@
+------------------------------------------
+The Econ-ARK infrastructure is built around creating a sustainable community with respect to various different use cases and the challenges of creating sustainable scientific software. We discuss some of the challenges of bridging work across user roles of Researchers, Publishers, Pedagogy and Software engineering. This is discussed from the point of view of economics research and tools in the Sc ...
+
+**Decoupling scientific content from code** A lot of scientific code is written as part of academic research projects where the incentives aren't closely aligned with those of creating scientific software (The recent case of UK COVID microsimulation code :cite:`covidsim2020`, brings out a stronger need of creating scientific software with the correct incentives). Initiatives like Journal of Open  ...
```
- Add a period after "code"
- `overlap of the researcher, publisher, and software engineer roles` (all singular)
- Please rephrase the last sentence and/or break into two sentences as the meaning is not clear.

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,625 @@
+  This decoupling exercise also helps with reproducibility part of research projects as it gives other researchers necessary tools to examine the research artifacts. The decoupling can also be extended to the data used in data-intensive research projects, which can also exist as a part of the scientific software library.
+
+
+**Reproducible builds of scientific content**, Reproducibility crisis has been plaguing academic research for some time and the current ecosystem of software packaging and distribution certainly doesn't help it. To tackle this in Econ-ARK we have used containerization technologies like Docker to solve this problem. Tools like Repo2Docker :cite:`repo2d` further help us with creating reproducible b ...
```
- Add a period after "content"
- `...still requires...` (add "s")
- `pre-built`
- `"one-click" (via commands)` (if I understand correctly, this means to say that artifacts are generated not by actual "clicks" of a button but via a command issued in a terminal, notebook, etc.)

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,625 @@
+software, as many computational models are not being presented
+in this rich interactive way.
+
+**Pedagogy Teaching resources**  To keep the wheels turning in a research discipline we require effective pedagogical resources, especially in domains which are increasingly using scientific software to further research. After creating pedagogical content we are faced with the next hard challenge of creating an effective teaching infrastructure. The crème de la crème of the SciPy community has fa ...
```
- `Pedagogy/Teaching resources.` (add "/" and period, or correct as necessary)
- `The reproducibility crisis...`
- `...your local environment. But luckily...`
- `...teaching courses that depend heavily on scientific software.`
- `...used MyBinder (publicly and privately hosted)...`
- `...especially for students who...`

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,625 @@
+
+**Interactive scientific publication**. The publication of the Econ-ARK based analysis of the
+consumption response to the CARES Act :cite:`carroll2020modeling`
+was accompanying by an online Dashboard [#]_ that allowed users
```
`accompanying` --> `accompanied`

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,625 @@
+knowledge to students, but to public policy makers and other
+economists.
+
+.. [#] \url{https://mybinder.org/v2/gh/econ-ark/Pandemic/master?urlpath=voila%2Frender%2FCode%2FPython%2Fdashboard.ipynb}
```
I don't think you need `\url` and curly brackets (check the rendered pdf)

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,625 @@
+However, researchers are not trained to create these Dashboards
+as they are trained to write research papers.
+This limits the scholarly impact that domains specific research
+software, as many computational models are not being presented
```
"impact that domains specific research software" <-- meaning unclear; impact that does what?

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,625 @@
+    print(calculate_MA(data, 'GOOG', 5))
+
+
+  Initial decisions like hard coding variables (which happens in a lot of academic research projects) in the code while creating the research artifact could lead away from creating a well defined reusable scientific software library. This seems trivial for people with a software engineering background but not necessarily for others. We discuss this further in our recommendation of software design ...
```
- `To tackle this...` (remove "and" and start a new sentence)
- `...and maintain the research artifacts...` (remove "we")

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,625 @@
+
+  Initial decisions like hard coding variables (which happens in a lot of academic research projects) in the code while creating the research artifact could lead away from creating a well defined reusable scientific software library. This seems trivial for people with a software engineering background but not necessarily for others. We discuss this further in our recommendation of software design ...
+  
+  This decoupling exercise also helps with reproducibility part of research projects as it gives other researchers necessary tools to examine the research artifacts. The decoupling can also be extended to the data used in data-intensive research projects, which can also exist as a part of the scientific software library.
```
- `...helps with the reproducibility part...` (add "the")
- `...gives other researchers the necessary tools...` (add "the")

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,625 @@
+core curriculum for graduate students.
+Some Economics departments already offer a course
+on Computational Methods, analogous to earlier courses
+on Mathematical Methods", Econometrics, or other methods
```
remove stray quotes

mepa reviewed (commented):
Sebastian and Mridul, I enjoyed reading your paper! My comments are minor - mainly requests for clarifications, typo corrections, etc.

sbenthall commented:
Thank you very much @mepa 
I'll make the necessary edits this week.

deniederhut commented:
Hi @sbenthall ! Have you had a chance to incorporate this review into your submission?

sbenthall committed:
edits from @mepa

sbenthall commented:
@deniederhut I was just able to get to this today. I've pushed the edits based on @mepa 's review.

A couple of the more substantive recommendations (the example of an ABM, the question about how the roles complicated the architecture) I've made notes about but may not be able to resolve quickly.

Will there be another review?

deniederhut commented:
> Will there be another review?

The expectation is that this is more of an ongoing process of collaboration with Meghann to get this paper ready for print by 2 July.

deniederhut commented:
Hi @mepa ! Are you happy with the updates to the paper?

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -28,10 +28,20 @@ Software Engineering as Research Method: Aligning Roles in Econ-ARK
+   learn computational and inferential thinking,
+   but are not equipped to participate in the full reproduction
+   of the scientific field unless they are also trained in
+   software engineering.
```
please check grammar of this sentence

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -51,12 +61,15 @@ While originally used as a tool for teaching mathematics,
+agent-based modeling (ABM) toolkit
 Mesa :cite:`masad2015mesa`.
 
+.. Example of ABM 
```
Thanks, I'll keep an eye out for this addition and review it if you get a chance to write up an example of ABM.

mepa replied:
💯

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -181,7 +194,7 @@ research into heterogeneous agent (HA)
 which became a research priority
 after the 2008 financial crisis revealed the weaknesses in the
-then-dominant representative agent (RA) based paradigm [#]
+then-dominant representative agent (RA) based paradigm [#]_
```
It renders correctly now! Please add a period to complete the sentence :)

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -442,17 +457,35 @@ or at another workplace.
+**Decoupling scientific content from code.** A lot of scientific code
+is written as part of academic research projects where the incentives
+aren't closely aligned with those of creating scientific software
+(The recent case of UK COVID microsimulation code :cite:`covidsim2020`,
```
remove comma

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -442,17 +457,35 @@ or at another workplace.
+aren't closely aligned with those of creating scientific software
+(The recent case of UK COVID microsimulation code :cite:`covidsim2020`,
+brings out a stronger need of creating scientific software with the
+correct incentives).
```
period inside parentheses

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -442,17 +457,35 @@ or at another workplace.
+(The recent case of UK COVID microsimulation code :cite:`covidsim2020`,
+brings out a stronger need of creating scientific software with the
+correct incentives).
+Initiatives like Journal of Open Source Software (JOSS) helps to align
```
Initiatives...help (remove 's')

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -442,17 +457,35 @@ or at another workplace.
+When scientific code written by researchers is geared towards the
+publishable end result like a paper, it can lead to short-sighted
+design choices that in a broader software context are known as
+"technical debt" :cite: `kruchten2012technical`. An illustraiont
```
- this isn't rendering correctly in the pdf (one can download and check from [here](http://procbuild.scipy.org/))
- `illustration`

Also, this sentence/example strikes a cord for me. I sometimes see the tendency in industry, also. Good software engineering practices (or an effort toward them), even in the research phase of a project, seem to enable more accurate and easier tech transfer of research artifacts, and more fluid collaboration between teammates.

sbenthall replied:
Yes!

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -512,17 +586,35 @@ substance of the model. Here, that paradigm is used not to convey
+and it is not hard to create a monster out of your local
+environment. But luckily tools like MyBinder and JupyterHub
+have drastically reduced the work required to set up a stable
+environment required for teaching courses that depend heavily
```
...heavily *on* scientific...

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -493,16 +526,57 @@ The Econ-ARK infrastructure is built around creating a sustainable community wit
+Econ-ARK based analysis of the
 consumption response to the CARES Act :cite:`carroll2020modeling`
-was accompanying by an online Dashboard [#]_ that allowed users
+was accompanied by an online Dashboard [#]_ that allowed users
```
allows? (present tense since the dashboard still exists)

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -493,16 +526,57 @@ The Econ-ARK infrastructure is built around creating a sustainable community wit
 consumption response to the CARES Act :cite:`carroll2020modeling`
-was accompanying by an online Dashboard [#]_ that allowed users
+was accompanied by an online Dashboard [#]_ that allowed users
 to change parameters of the model and visualized their impact
```
visualize?

mepa reviewed (commented):
Thanks for the changes, @sbenthall. Sorry I'm just getting around to addressing them. 

I've added a few more comments in this second review, and plan to read the pdf end-to-end tomorrow evening to check for potential final changes. It's looking good!

sbenthall committed:
edits

sbenthall commented:
Hello.

Thanks for these additional edits, @mepa.

I've addressed your comments. I also did my own editing pass for punctuation, style, etc. These are all in the latest commit.

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,722 @@
+The Econ-ARK project :cite:`carroll2018econ`
+is a toolkit for the structural
+modeling of optimizing economic choices by heterogenous agents.
+A primary goal of its flagship software library, HARK
```
remove comma

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,722 @@
+**Example**. A paradigmatic, simple example of the kind
+of problem studied using Econ-ARK is the microeconomic
+dynamic stochastic optimization problem of calculating
+the mathematically optimal amount to save. :cite:`carroll2011solution`
```
period after citation

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,722 @@
+As a consequence there is a skills gap: researchers often
+have programming ability, but not the software engineering
+and IT training that is necessary to fully realize the
+vision of the software's potential. :cite:`crouch2013software`
```
period after citation

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,722 @@
+loosely on the Journal of Open Source Software (JOSS),
+which gives academic publication credit to the creators
+of scientific software tools.
+Preliminery efforts towards such a journal have been
```
Preliminary

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,722 @@
+This decoupling exercise also helps with the reproducibility of
+research projects as it gives other researchers the necessary
+tools to examine the research artifacts. The decoupling can
+also enables the use of empirical data and model fitting techniques,
```
enable

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,722 @@
+
+We are definitely not the first push for more training to
+scientific researchers about general software design and best
+practices (software versioning, CI, testing).
```
spell out "continuous integration"?

mepa commented on papers/sebastian_benthall/sebastian_benthall.rst:
```diff
@@ -0,0 +1,722 @@
+Researchers writing code with knowledge about software design will have
+more success in creating a sustainable community.
+Our contribution in this paper is to discuss how software design can be
+reconceived as a scientific method, as opposed to a peripheral skill.
```
great punch line! :)

mepa commented:
LGTM, @sbenthall. I really enjoyed reading this paper! I added a few minor edit comments as I read through it one last time, but I'm happy with the paper as is.

FYI, @deniederhut

sbenthall commented:
Thanks again @mepa 

@deniederhut may I make the edits suggested by @mepa even though it's past the deadline?

mepa commented:
> @deniederhut may I make the edits suggested by @mepa even though it's past the deadline?

Yes, please feel free to make the changes today (tomorrow at the latest), @sbenthall. Thanks!

deniederhut commented on papers/sebastian_benthall/mybib.bib:
```diff
@@ -0,0 +1,461 @@
+@misc{adhikari2017computational,
+  title={Computational and Inferential Thinking: The Foundations of Data Science},
+  author={Adhikari, Ani and DeNero, John},
+  year={2017}
```
Looks like this is missing a url

deniederhut commented on papers/sebastian_benthall/mybib.bib:
```diff
@@ -0,0 +1,461 @@
+@article{barba2016computational,
+  title={Computational Thinking: I do not think it means what you think it means},
+  author={Barba, Lorena A and others},
+  journal={Lorena A. Barba Group},
```
I'm not used to seeing a blog post referenced as a journal. Was this meant to be a link to the website?

sbenthall replied:
I think for this and the other blog post reference, I used a citation generator that uses the 'journal' field to refer to the publication venue, which in this case is the blog title.

I believe this is common practice? Though there are other errors in this reference, and there should be a URL

deniederhut commented on papers/sebastian_benthall/mybib.bib:
```diff
@@ -0,0 +1,461 @@
+  year={2012}
+}
+
+@misc{benthall_2019, title={US Research Software Sustainability Institute}, url={http://urssi.us/blog/2019/02/25/software-incubator-workshop-a-synthesis/}, journal={US Research Software Sustainability Institute}, author={Benthall, Sebastian}, year={2019}, month={Feb}}
```
Is this a journal or a blog?

deniederhut commented on papers/sebastian_benthall/mybib.bib:
```diff
@@ -0,0 +1,461 @@
+@misc{benthall_2019, title={US Research Software Sustainability Institute}, url={http://urssi.us/blog/2019/02/25/software-incubator-workshop-a-synthesis/}, journal={US Research Software Sustainability Institute}, author={Benthall, Sebastian}, year={2019}, month={Feb}}
+
+@article{blank2019nbgrader,
+  title={nbgrader: A tool for creating and grading assignments in the Jupyter Notebook},
```
This is missing a DOI

deniederhut commented on papers/sebastian_benthall/mybib.bib:
```diff
@@ -0,0 +1,461 @@
+}
+
+@article{cock2009biopython,
+  title={Biopython: freely available Python tools for computational molecular biology and bioinformatics},
```
This is also missing a DOI

deniederhut commented on papers/sebastian_benthall/mybib.bib:
```diff
@@ -0,0 +1,461 @@
+}
+
+@article{carroll2018econ,
+  title={The Econ-ARK and HARK: Open Source Tools for Computational Economics},
```
Was this the SciPy paper? If so, this should be an in-proceedings, with proceedings title, editors, doi, etc.

deniederhut reviewed (changes requested):
Hey @sbenthall ! Sorry to to bring this up so late in the process, but it looks like you didn't finish the reference section?

sbenthall commented:
Yikes! Sorry--I took this July 4th weekend off.
I'll work on this now.

sbenthall committed:
edits: mepa last comments; biblio fixes from dniederhut

sbenthall commented:
Edits are in.

deniederhut commented on papers/sebastian_benthall/mybib.bib:
```diff
@@ -0,0 +1,479 @@
+}
+
+@article{breiman2001statistical,
+  title={Statistical modeling: The two cultures (with comments and a rejoinder by the author)},
```
I think this one still needs a DOI

MridulS replied:
for ref: doi:10.1214/ss/1009213726

deniederhut replied:
Did we add this?

deniederhut commented on papers/sebastian_benthall/mybib.bib:
```diff
@@ -0,0 +1,479 @@
+
+@article{carroll2018hands,
+  title={Hands-On Heterogeneous Agent Macroeconomics},
+  author={Carroll, Christopher D and White, Matthew N},
```
is this missing a journal?

MridulS replied:
This is a reference to a course, https://safe-frankfurt.de/fileadmin/user_upload/editor_common/Events/Chris_Carrol_Syllabus.pdf . What should we put in here?

deniederhut replied:
Okay. I think I'm thrown off by seeing all of these listed as articles. Would it make sense to use `@misc` for these kinds of bibfile entries?

MridulS replied:
Sure `@misc`= sounds good.

deniederhut commented on papers/sebastian_benthall/mybib.bib:
```diff
@@ -0,0 +1,479 @@
+
+@article{carroll2020modeling,
+  title={Modeling the Consumption Response to the CARES Act},
+  author={Carroll, Christopher D and Crawley, Edmund and Slacalek, Jiri and White, Matthew N},
```
is this also missing a journal?

MridulS replied:
This isn't published yet, it's in the review process.

deniederhut replied:
That's okay. Could we add the journal name and something like `in review` for the year?

deniederhut commented on papers/sebastian_benthall/mybib.bib:
```diff
@@ -0,0 +1,479 @@
+@article{crouch2013software,
+  title={The Software Sustainability Institute: changing research software attitudes and practices},
+  author={Crouch, Stephen and Hong, Neil Chue and Hettrick, Simon and Jackson, Mike and Pawlik, Aleksandra and Sufi, Shoaib and Carr, Les and De Roure, David and Goble, Carole and Parsons, Mark},
+  journal={Computing in Science \& Engineering},
```
I think this one is also missing a DOI

MridulS replied:
for ref: doi.org/10.1109/MCSE.2013.133

deniederhut commented on papers/sebastian_benthall/mybib.bib:
```diff
@@ -0,0 +1,479 @@
+
+@article{dalle2004toward,
+  title={Toward collaborative open-source technology transfer},
+  author={Dalle, Jean-Michel and Rousseau, Guillaume},
```
I think this too is missing a DOI

MridulS replied:
for ref: doi:10.1049/ic:20040262

Commit:
Update mybib.bib

deniederhut commented:
Also, it looks like the editor names and some other info is missing in the rendered pdf. Our expectation is that they will look something like this:

<img width="278" alt="Screenshot 2020-07-06 19 22 59" src="https://user-images.githubusercontent.com/6868396/86674448-504aeb00-bfbe-11ea-89b2-5412ed5bbea3.png">

Commit:
Update mybib.bib

sbenthall committed:
Merge pull request #14 from MridulS/patch-2

Update mybib.bib

sbenthall commented:
@deniederhut can you point to an example of a scipy paper that has it's reference to another scipy paper correctly bibtexed?

deniederhut commented:
Sure, the screencap is from the rendered pdf of https://github.com/scipy-conference/scipy_proceedings/pull/570/files#diff-543a34787bd14e4ee6f7aa23dcdd1226R177

sbenthall commented:
Perfect. Thank you! I'll correct all those references now. Thanks for your patience.

sbenthall commented:
I see that BibTex marks up the capitalization of titles.
Is that expected?

sbenthall commented:
Ah! The proceedings website generates the correct BibTex. fantastic.

sbenthall committed:
bibtex fixes; better scipy conf references

sbenthall committed:
Merge branch '2020_econ_ark' of github.com:sbenthall/scipy_proceedings into 2020_econ_ark

sbenthall commented:
Should be good now. sorry for the delay

deniederhut commented on papers/sebastian_benthall/mybib.bib:
```diff
@@ -0,0 +1,490 @@
+
+@article{donoho201750,
+  title={50 years of data science},
+  author={Donoho, David},
```
I think this is missing a DOI

deniederhut commented on papers/sebastian_benthall/mybib.bib:
```diff
@@ -0,0 +1,490 @@
+
+@article{guzdial2008education,
+  title={Education Paving the way for computational thinking},
+  author={Guzdial, Mark},
```
I think this is also missing a DOI

sbenthall replied:
There was indeed a DOI for this. I've added it now.

deniederhut commented on papers/sebastian_benthall/mybib.bib:
```diff
@@ -0,0 +1,490 @@
+
+@inproceedings{hamrick2016creating,
+  title={Creating and grading IPython/Jupyter notebook assignments with NbGrader},
+  author={Hamrick, Jessica B},
```
May also be missing a DOI

deniederhut commented:
No worries! I think the references are still missing some DOIs though.

sbenthall committed:
adding many dois

sbenthall commented:
I've now gone through all the references, looked them up on CrossRef, and added a DOI if there wasn't one.

sbenthall committed:
adding doi for Guzdial article

deniederhut commented:
Sweet - thanks for going through all the references!

sbenthall commented:
Thanks so much, @deniederhut !!

sbenthall commented:
May I ask how often the proceedings site gets updated?
http://conference.scipy.org/proceedings/scipy2020/
